### DISCUSSION TITLE
Selectors

### DISCUSSION THREAD

**Discussion by ThatXliner** [lark-parser/lark#831] [Ideas]
Kind of like your other project [plyplus](https://github.com/erezsh/plyplus), I think Lark might benefit from CSS-ish queries and selectors. Because, AFIAK, writing a transformer/visitor is the only way to interact with the tree officially.

**Comment by erezsh** [comment]
Can you give me examples of situations or tasks which would be best done with selectors?

My experience was that they weren't that useful, and most of the time could be replaced with `find_data` or `scan_values`.

**Reply by ThatXliner** [reply]
Wait, what's find_data? Am I missing something? I probably skimmed over it....

**Reply by ThatXliner** [reply]
Ok, so I found it [in the docs](https://lark-parser.readthedocs.io/en/latest/classes.html#lark.Tree.find_data). What's the input `data` supposed to be (I.e. type of object)

**Reply by erezsh** [reply]
In the case of Lark, `data` is usually the name of rule that was matched. But it can be any object you put there (for example, by using a transformer)

**Reply by ThatXliner** [reply]
Now that I think of it, I guess using find_data could be used in most areas when you need to find a Tree or Token. But you would need to construct the Token or Tree for input.

But CSS also allows selectors that have look-behinds. Like this:

```css
element_that_should_exist_in_front > item.class {…}
```
It's probably a dumb suggestion, though.

AFIAK, this kind of look behind isn't possible.

**Reply by erezsh** [reply]
You should also be aware of `find_pred` and `scan_values`.

Definitely not a dumb suggestion, I implemented it myself for plyplus because I thought it would be useful.

My experience is that it isn't. There are things selectors can do that a simple search doesn't, but turns out that it very rarely occurs in parse-trees. Usually, transformers do the trick.

For the example you gave, you can do this:

```python
[item 
 for node in tree.find_data('element_that_should_exist_in_front')
 for item in node.find_data('item')
 if item.class = "class"
]
```

It's not bad enough to warrant a complicated new feature. And it definitely doesn't happen often enough.

**Reply by ThatXliner** [reply]
I guess so. I'll come back to this idea when I need it (I.e. *very* specific data-grabbing)

**Reply by ThatXliner** [reply]
The `tree.find_data` doesn't seem to have a recursive option... is that correct?

**Reply by MegaIng** [reply]
What do you mean with that?

**Reply by ThatXliner** [reply]
Basically, `tree.find_data` doesn't have an option to recursively search for `data`. In other words, it won't search in the node's trees's tree's.....

---

```py
Tree('foo', [Tree('inner_foo', [Tree('inner_foo2', [Token('STRING', 'bar')]), Token('STRING', 'baz')])
```
With that tree, `foo.find_data` will not be able to find "bar".

**Reply by MegaIng** [reply]
It can't find `Token`s, no. But it does recursively search subtrees.

**Reply by ThatXliner** [reply]
I see. So I'll need to apply a transformer to the tree to, I don't know, turn the `Token`s into `str`ings and *then* do `tree.find_data`?

**Reply by MegaIng** [reply]
`Tree.find_data` finds `Tree` instance with a `data` attribute that matches the request one. It ignores all other children.

**Reply by ThatXliner** [reply]
Oh ok. But finding tokens would be nice. There's no issue for that, right?

**Reply by MegaIng** [reply]
.... That is what `scan_values` is there for.

**Reply by ThatXliner** [reply]
Huh? `lark.Tree` doesn't have a `scan_values`. At least in the docs...
<img width="723" alt="Screen Shot 2021-03-27 at 3 04 20 PM" src="https://user-images.githubusercontent.com/66848002/112736153-b7724c00-8f0d-11eb-97c8-4c67e606dde2.png">

**Reply by erezsh** [reply]
@ThatXliner You're right, it's not documented. I will add it now

**Reply by erezsh** [reply]
@ThatXliner Added: https://lark-parser.readthedocs.io/en/latest/classes.html#lark.Tree.scan_values
